src/goto-instrument/synthesizer/expr_enumerator.h

```diff
@@ -0,0 +1,511 @@
+  }
+
+  /// Given a list `enumerators` of enumerators, return the caresian product
+  /// of expressions enumerated by each enumerator in the list.
+  std::list<exprst> cartesian_product_of_enumerators(
+    const enumeratorst &enumerators,
+    unsigned int pos = 0);
+
+  /// Enumerate all partitions of n into k components.
+  /// Order of partitions is considered relevant.
+  std::list<partitiont> get_partitions(unsigned int n, unsigned int k);
+
+  exprst enumerate() override
+  {
+    exprst result = exprst();
+
+    // Enumerate nothing when `expected_size` is too small to be partitioned.
+    if(expected_size - 1 < arity)
+      return result;
+
+    // For every possible partition, set `expected_size` of
+    // each sub-enumerator to be the corresponding component in the partition.
+    for(const auto &partition : get_partitions(expected_size - 1, arity))
+    {
+      if(!is_good_partition(partition))
+        continue;
+
+      set_sub_enumerator_sizes(partition);
+
+      // Compute the cartesian product as result.
+      for(const auto &product_tuple :
+          cartesian_product_of_enumerators(sub_enumerators))
+      {
+        // Optimization: rule out equivalent expressions
+        // using certain equivalence class.
+        // Keep only representation tuple of each equivalence class.
+        if(is_equivalence_class_representation(product_tuple))
+          result.push_back(instantiate(product_tuple));
+      }
+    }
+
+    return result;
+  }
+
+  void set_size(unsigned int new_size) override
+  {
+    expected_size = new_size;
+  }
+
+  non_leaf_enumeratort *clone() const override
+  {
+    return new non_leaf_enumeratort(sub_enumerators, is_none_enumerator);
+  }
+
+  /// As default, keep all expression tuples.
+  virtual bool is_equivalence_class_representation(const exprst &es)
+  {
+    return true;
+  }
+
+protected:
+  /// Propogate partition to sub-enumerators.
+  void set_sub_enumerator_sizes(const partitiont &sizes)
+  {
+    INVARIANT(sizes.size() == arity, "arity not equal to number of arguments");
+    auto front = sizes.begin();
+    for(auto &sub_enumerator : sub_enumerators)
+    {
+      sub_enumerator->set_size(*front);
+      std::advance(front, 1);
+    }
+  }
+
+  /// Instantiate a list of expressions to a expression using
+  /// a top-level operator.
+  virtual exprt instantiate(const exprst &exprs)
+  {
+    return true_exprt();
+  }
+
+  unsigned int arity;
+  unsigned int expected_size;
+  enumeratorst sub_enumerators;
+  std::function<bool(const partitiont &)> is_good_partition;
+};
+
+/// Enumerator that enumerates binary functional expressions.
+class binary_functional_enumeratort : public non_leaf_enumeratort
+{
+public:
+  binary_functional_enumeratort(
+    const irep_idt &op,
+    const enumerator_baset &enumerator_1,
+    const enumerator_baset &enumerator_2,
+    std::function<bool(const partitiont &)> partition_check)
+    : op_id(op)
+  {
+    is_none_enumerator = false;
```

Comment:
Redundant? But, really, should be explicitly invoking the parent constructor and get rid of those default parameters.